File: src/types/modpack.ts
export type ModpackFormat = 'curseforge' | 'modrinth' | 'ftb' | 'technic' | 'bonjour' | 'unknown'

export type ModEntrySource = 'curseforge' | 'modrinth' | 'direct' | 'local'
export type ConfigEntrySource = 'override' | 'embedded'
export type TestCheckCategory = 'dependency' | 'conflict' | 'resource' | 'compatibility' | 'performance'
export type TestCheckStatus = 'pass' | 'warn' | 'fail'
export type ForkConflictType = 'mod_added_upstream' | 'mod_removed_upstream' | 'mod_version_conflict' | 'config_conflict'
export type ForkConflictResolution = 'keep_fork' | 'keep_upstream' | 'merge'
export type SyncRoomStatus = 'waiting' | 'syncing' | 'complete'
export type SyncParticipantStatus = 'waiting' | 'downloading' | 'installing' | 'complete' | 'error'
export type ConfigChangeType = 'added' | 'modified' | 'removed'

export interface ModpackEntry {
  id: string
  name: string
  version: string
  author: string
  description: string
  gameVersion: string
  modLoader?: string
  modLoaderVersion?: string
  format: ModpackFormat
  sourcePath?: string
  instanceId?: string
  instanceName?: string
  installedAt: number
  lastUpdatedAt?: number
  mods: ModpackModEntry[]
  configs: ModpackConfigEntry[]
  overridesDir?: string
  iconUrl?: string
  sourceUrl?: string
  isFork?: boolean
  forkInfo?: ModpackForkInfo
}

export interface ModpackForkInfo {
  originalPackId: string
  originalPackName: string
  originalVersion: string
  forkCreatedAt: number
  addedMods: string[]
  removedMods: string[]
  modifiedConfigs: string[]
  lastSyncedVersion: string
  hasUpstreamUpdate: boolean
}

export interface ModpackManifest {
  format: ModpackFormat
  name: string
  version: string
  author: string
  description: string
  gameVersion: string
  modLoader?: string
  modLoaderVersion?: string
  mods: ModpackModEntry[]
  configs: ModpackConfigEntry[]
  overridesDir?: string
  sourceUrl?: string
  iconUrl?: string
}

export interface ModpackModEntry {
  fileName: string
  projectId?: number
  fileId?: number
  downloadUrl?: string
  hash?: string
  size?: number
  source: ModEntrySource
  required: boolean
  folderPath?: string
}

export interface ModpackConfigEntry {
  relativePath: string
  source: ConfigEntrySource
  hash?: string
  size?: number
}

export interface ModpackInstallProgress {
  phase: 'parsing' | 'downloading' | 'configuring' | 'verifying' | 'complete'
  current: number
  total: number
  currentItem: string
  bytesDownloaded: number
  bytesTotal: number
}

export interface ModpackInstallResult {
  success: boolean
  instanceId?: string
  instanceName: string
  modsInstalled: number
  configsRestored: number
  errors: string[]
}

export interface ModpackCreateSession {
  id: string
  name: string
  version: string
  author: string
  description: string
  gameVersion: string
  modLoader: string
  modLoaderVersion: string
  mods: ModpackCreateMod[]
  configs: ModpackCreateConfig[]
  overrides: string[]
  sourceInstanceId?: string
  createdAt: number
}

export interface ModpackCreateMod {
  fileName: string
  filePath: string
  source: ModEntrySource
  projectId?: number
  fileId?: number
  downloadUrl?: string
  hash?: string
  required: boolean
}

export interface ModpackCreateConfig {
  relativePath: string
  content: string
  included: boolean
}

export interface ModpackDiff {
  added: ModpackModEntry[]
  removed: ModpackModEntry[]
  updated: ModpackModEntry[]
  configChanges: ModpackConfigChange[]
  summary: string
}

export interface ModpackConfigChange {
  path: string
  changeType: ConfigChangeType
}

export interface ModpackUpdateResult {
  success: boolean
  modsAdded: number
  modsRemoved: number
  modsUpdated: number
  configChanges: number
  errors: string[]
  rolledBack: boolean
}

export interface ModpackForkMergeResult {
  success: boolean
  conflicts: ModpackForkConflict[]
  autoResolved: number
  manualRequired: number
}

export interface ModpackForkConflict {
  modFileName: string
  conflictType: ForkConflictType
  upstreamAction: string
  forkAction: string
  resolution?: ForkConflictResolution
}

export interface ModpackRating {
  modpackId: string
  completeness: number
  stability: number
  performance: number
  difficulty: number
  innovation: number
  overall: number
  reviewCount: number
}

export interface ModpackReview {
  id: string
  modpackId: string
  userId: string
  username: string
  rating: number
  completeness: number
  stability: number
  performance: number
  difficulty: number
  innovation: number
  content: string
  createdAt: number
  likes: number
}

export interface ModpackRecommendation {
  modpackId: string
  name: string
  score: number
  reason: string
  tags: string[]
}

export interface ModpackTestResult {
  passed: boolean
  checks: ModpackTestCheck[]
  overallScore: number
  estimatedStartupTime: number
  estimatedFps: number
  warnings: string[]
}

export interface ModpackTestCheck {
  id: string
  name: string
  category: TestCheckCategory
  status: TestCheckStatus
  message: string
  details?: string
}

export interface ModpackSyncRoom {
  id: string
  code: string
  hostId: string
  hostName: string
  modpackName: string
  modpackVersion: string
  gameVersion: string
  modLoader: string
  modCount: number
  participants: ModpackSyncParticipant[]
  createdAt: number
  status: SyncRoomStatus
}

export interface ModpackSyncParticipant {
  id: string
  name: string
  status: SyncParticipantStatus
  progress: number
  joinedAt: number
}

export interface ModpackPerformanceBenchmark {
  modpackId: string
  modpackName: string
  modCount: number
  minRam: number
  recommendedRam: number
  startupTimeMin: number
  startupTimeMax: number
  fpsMin: number
  fpsAvg: number
  fpsMax: number
  testConfig: string
  sampleCount: number
  lastUpdated: number
}

export function createEmptyModpackCreateSession(): ModpackCreateSession {
  return {
    id: `mpc-${Date.now()}-${Math.random().toString(36).substr(2, 9)}`,
    name: '',
    version: '1.0.0',
    author: '',
    description: '',
    gameVersion: '',
    modLoader: 'forge',
    modLoaderVersion: '',
    mods: [],
    configs: [],
    overrides: [],
    createdAt: Date.now(),
  }
}

File: bonjour-minecraft-launcher/src-tauri/src/models/modpack.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackEntry {
    pub id: String,
    pub name: String,
    pub version: String,
    pub author: String,
    pub description: String,
    pub game_version: String,
    pub mod_loader: Option<String>,
    pub mod_loader_version: Option<String>,
    pub format: ModpackFormat,
    pub source_path: Option<String>,
    pub instance_id: Option<String>,
    pub instance_name: Option<String>,
    pub installed_at: i64,
    pub last_updated: Option<i64>,
    pub mods: Vec<ModpackModEntry>,
    pub configs: Vec<ModpackConfigEntry>,
    pub overrides_dir: Option<String>,
    pub icon_url: Option<String>,
    pub source_url: Option<String>,
    pub is_fork: Option<bool>,
    pub fork_info: Option<ModpackForkInfo>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ModpackFormat {
    CurseForge,
    Modrinth,
    Ftb,
    Technic,
    Bonjour,
    Unknown,
}

impl Default for ModpackFormat {
    fn default() -> Self {
        ModpackFormat::Unknown
    }
}

impl ModpackFormat {
    pub fn as_str(&self) -> &str {
        match self {
            ModpackFormat::CurseForge => "curseforge",
            ModpackFormat::Modrinth => "modrinth",
            ModpackFormat::Ftb => "ftb",
            ModpackFormat::Technic => "technic",
            ModpackFormat::Bonjour => "bonjour",
            ModpackFormat::Unknown => "unknown",
        }
    }

    pub fn from_str(s: &str) -> Self {
        match s.to_lowercase().as_str() {
            "curseforge" => ModpackFormat::CurseForge,
            "modrinth" => ModpackFormat::Modrinth,
            "ftb" => ModpackFormat::Ftb,
            "technic" => ModpackFormat::Technic,
            "bonjour" => ModpackFormat::Bonjour,
            _ => ModpackFormat::Unknown,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackModEntry {
    pub file_name: String,
    pub project_id: Option<i64>,
    pub file_id: Option<i64>,
    pub download_url: Option<String>,
    pub hash: Option<String>,
    pub size: Option<u64>,
    pub source: ModEntrySource,
    pub required: bool,
    pub folder_path: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ModEntrySource {
    CurseForge,
    Modrinth,
    Direct,
    Local,
}

impl Default for ModEntrySource {
    fn default() -> Self {
        ModEntrySource::Local
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackConfigEntry {
    pub relative_path: String,
    pub source: ConfigEntrySource,
    pub hash: Option<String>,
    pub size: Option<u64>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ConfigEntrySource {
    Override,
    Embedded,
}

impl Default for ConfigEntrySource {
    fn default() -> Self {
        ConfigEntrySource::Override
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackForkInfo {
    pub original_pack_id: String,
    pub original_pack_name: String,
    pub original_version: String,
    pub fork_created_at: i64,
    pub added_mods: Vec<String>,
    pub removed_mods: Vec<String>,
    pub modified_configs: Vec<String>,
    pub last_synced_version: String,
    pub has_upstream_update: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackDiff {
    pub added: Vec<ModpackModEntry>,
    pub removed: Vec<ModpackModEntry>,
    pub updated: Vec<ModpackModEntry>,
    pub config_changes: Vec<ModpackConfigChange>,
    pub summary: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackConfigChange {
    pub path: String,
    pub change_type: ConfigChangeType,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ConfigChangeType {
    Added,
    Modified,
    Removed,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackUpdateResult {
    pub success: bool,
    pub mods_added: u32,
    pub mods_removed: u32,
    pub mods_updated: u32,
    pub config_changes: u32,
    pub errors: Vec<String>,
    pub rolled_back: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackTestResult {
    pub passed: bool,
    pub checks: Vec<ModpackTestCheck>,
    pub overall_score: u32,
    pub estimated_startup_time: u32,
    pub estimated_fps: u32,
    pub warnings: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackTestCheck {
    pub id: String,
    pub name: String,
    pub category: TestCheckCategory,
    pub status: TestCheckStatus,
    pub message: String,
    pub details: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TestCheckCategory {
    Dependency,
    Conflict,
    Resource,
    Compatibility,
    Performance,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TestCheckStatus {
    Pass,
    Warn,
    Fail,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackPerformanceBenchmark {
    pub modpack_id: String,
    pub modpack_name: String,
    pub mod_count: u32,
    pub min_ram: u64,
    pub recommended_ram: u64,
    pub startup_time_min: u32,
    pub startup_time_max: u32,
    pub fps_min: u32,
    pub fps_avg: u32,
    pub fps_max: u32,
    pub test_config: String,
    pub sample_count: u32,
    pub last_updated: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackSyncRoom {
    pub id: String,
    pub code: String,
    pub host_id: String,
    pub host_name: String,
    pub modpack_name: String,
    pub modpack_version: String,
    pub game_version: String,
    pub mod_loader: String,
    pub mod_count: u32,
    pub participants: Vec<ModpackSyncParticipant>,
    pub created_at: i64,
    pub status: SyncRoomStatus,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SyncRoomStatus {
    Waiting,
    Syncing,
    Complete,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackSyncParticipant {
    pub id: String,
    pub name: String,
    pub status: SyncParticipantStatus,
    pub progress: u32,
    pub joined_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum SyncParticipantStatus {
    Waiting,
    Downloading,
    Installing,
    Complete,
    Error,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackRecommendation {
    pub modpack_id: String,
    pub name: String,
    pub score: f64,
    pub reason: String,
    pub tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackRating {
    pub modpack_id: String,
    pub completeness: u32,
    pub stability: u32,
    pub performance: u32,
    pub difficulty: u32,
    pub innovation: u32,
    pub overall: u32,
    pub review_count: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackForkConflict {
    pub mod_file_name: String,
    pub conflict_type: ForkConflictType,
    pub upstream_action: String,
    pub fork_action: String,
    pub resolution: Option<ForkConflictResolution>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ForkConflictType {
    ModAddedUpstream,
    ModRemovedUpstream,
    ModVersionConflict,
    ConfigConflict,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum ForkConflictResolution {
    KeepFork,
    KeepUpstream,
    Merge,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ModpackForkMergeResult {
    pub success: bool,
    pub conflicts: Vec<ModpackForkConflict>,
    pub auto_resolved: u32,
    pub manual_required: u32,
}
